lxd/instances_post.go

```diff
@@ -1137,18 +1084,7 @@ func instancesPost(d *Daemon, r *http.Request) response.Response {
 
 		// If no target member was selected yet, pick the member with the least number of instances.
 		if targetMemberInfo == nil {
-			err = s.DB.Cluster.Transaction(r.Context(), func(ctx context.Context, tx *db.ClusterTx) error {
-				targetMemberInfo, err = tx.GetNodeWithLeastInstances(ctx, candidateMembers)
-				if err != nil {
-					return err
-				}
-
-				if targetMemberInfo == nil {
-					return api.StatusErrorf(http.StatusBadRequest, "No suitable cluster member could be found")
```

Comment:
Lets lose the `getNodeWithLeastInstances` function, as we are trying to move away from creating transactions inside lower level functions (for the reasons above that I described around doing as much as possible inside fewer transactions) so it helps to keep the transactions being created visible.

Instead lets push down the `return api.StatusErrorf(http.StatusBadRequest, "No suitable cluster member could be found")` error into the `GetNodeWithLeastInstances` function itself, so that `GetNodeWithLeastInstances` can never return a nil node info without a non-nil error.

We can then update the call sites of `GetNodeWithLeastInstances` to directly return the error or if needed to detect the specific not found error using `api.StatusErrorCheck(err, http.StatusNotFound)`

Reply:
The logic is now moved into `GetNodeWithLeastInstances`. I have intentionally tried to keep the first `.Transaction()` as small as possible, breaking out for other tasks and then starting a new one if this step is required (still no member found).